radvd.8.man
.\"
.\"   $Id: radvd.8.man,v 1.1.1.1 1997/10/14 17:17:40 lf Exp $
.\"
.\"   Authors:
.\"    Lars Fenneberg		<lf@elemental.net>	 
.\"
.\"   This software is Copyright 1996 by the above mentioned author(s), 
.\"   All Rights Reserved.
.\"
.\"   The license which is distributed with this software in the file COPYRIGHT
.\"   applies to this software. If your distribution is missing this file, you
.\"   may request it from <lf@elemental.net>.
.\"
.\"
.\"
.TH RADVD 8 "8 Dec 1996" "radvd @VERSION@" ""
.SH NAME
radvd \- router advertisement daemon for IPv6
.SH SYNOPSIS
.B radvd
.B "[ \-vh ]"
.BI "[ \-d " debuglevel " ]"
.BI "[ \-C " configfile " ]"
.BI "[ \-m " method " ]"
.BI "[ \-l " logfile " ]"
.BI "[ \-f " facility " ]"

.SH DESCRIPTION
.B radvd
is the router advertisement daemon for IPv6. It listens to router
solicitations and sends router advertisements as described in
"Neighbor Discovery for IP Version 6 (IPv6)" (RFC 1970). With this
advertisements hosts can automatically configure there addresses and
some other parameters. They also can choose a default router based
on these advertisements.

.SH OPTIONS

For every one character option there is also a long option, which
is listed right next to the "short" option name:

.TP
.BR "\-v" , " \-\-version"
Displays the version of
.I radvd
and then aborts.
.TP
.BR "\-h" , " \-\-help"
Displays a short usage description and then aborts.
.TP
.BR "\-d " debuglevel, " \-\-debug " debuglevel
With this option you turn on debugging information. The debugging level is
an integer in the range from 1 to 4, from  quiet to very verbose. A
debugging level of 0 completely turns of debugging. If a debugging level 
greater than 0 is used,
.I radvd
doesn't background itself on start. The default debugging level is 0.
.TP
.BR "\-C " configfile, " \-\-config " configfile
Specifies an alternate config file. Normally the compiled in default
.I @PATH_RADVD_CONF@
is used.
.TP
.BR "\-m " method, " \-\-method " method
Specifies the logging method to use. Possibly values are:
.RS
.TP
.B none
Completely disables any logging.
.TP
.B logfile
Logs to the logfile which is specified by the
.I \-l
option. If no logfile is specified on the command line, then
a compiled in default is used (see next option).
.TP
.B stderr
Logs to standard error.
.TP
.B syslog
logs to syslog (default method).
.RE
.PP
.TP
.BR "\-l " logfile, " \-\-logfile " logfile
Specifies the logfile to use when using the logging method
.IR logfile .
The default logfile is
.IR @PATH_RADVD_LOG@ .
.TP
.BR "\-f " facility, " \-\-facility " facility
Specifies the facility (as an integer) when using syslog logging. Default
is @LOG_FACILITY@.
.SH FILES

.nf
@sbindir@/radvd
@PATH_RADVD_CONF@
@PATH_RADVD_LOG@
.fi
.SH BUGS

There certainly are some bugs. If you find them or have other
suggestions please contact Lars Fennberg <lf@elemental.net>.

.SH "SEE ALSO"

.BR radvd.conf (5)
.SH AUTHORS

.nf
Pedro Roque	<roque@di.fc.ul.pt> - wrote first version for Linux
Lars Fenneberg	<lf@elemental.net>  - current maintainer
Craig Metz	<cmetz@inner.net>   - port to NRL's IPv6 code for BSD4.4
.fi


CHANGES
$Id: CHANGES,v 1.1.1.1 1997/10/14 17:17:40 lf Exp $

09/14/97	EUI-64 "support"

09/01/97	centralized all includes into includes.h for easier
		maintainance

08/25/97	changed all occurences of IPV6_TXINFO/IPV6_RXINFO
		to IPV6_PKTINFO

08/25/97	removed some compatibility stuff for older Linux kernels

03/17/97	changed linux/if_arp.h to net/if_arp.h in device-linux.c
		and configure.in (for glibc compatibility)

03/17/97	integrated patch from Philip Blundell <pjb27@cam.ac.uk>
		into configure.in to check for inet_* in libc first and
		then for libinet6 (for glibc compatibility).

03/16/97	corrected small error in radvd.conf(5) man page and other
		small changes

03/16/97	removed linklocal.c source file from the distribution
		because it wasn't used anymore

03/16/97	added html version of the INTRO, removed ascii version.

02/12/97	missing directories are now created on make install

01/19/97	now also the rest of a line followed by a '#' is treated
		as a comment in the config file

01/16/97	added IPPROTO_IPV6/IPV6_HOPLIMIT support, so now incoming
		RS' and RAs are checked if there hoplimit is 255 (on Linux
		this only works with kernel version 2.1.21 and above)

01/15/97	lots of portability cleanups; radvd is now using autoconf
		for automatic configuration; removed support for the
		in. prefix

01/05/97	integrated BSD 4.4 patches from Craig <cmetz@inner.net>

12/21/96	release of version 0.3

12/15/96	the AdvSendAdvert flag works now

12/14/96	added radvdump, dumps all received RAs in a human
		readable form	

12/13/96	rewrote radvd to use new advanced API headers

12/08/96	added radvd.conf(5) man page

12/03/96	changed RAW_CHECKSUM to IPV6_CHECKSUM for kernels
		2.1.14 and above

11/24/96	added checking of received RAs according to RFC 1970

11/20/96	rewrote the default handling, now it should conform
		to RFC 1970

11/17/96	release of version 0.2
	
11/17/96	added -h (--help) and -f (--facility) option

11/17/96	added setsockopt(..., SOL_IPV6, IPV6_MULTICAST_HOPS), so
	 	that radvd works for kernel versions >2.1.7 (without this
	 	you get "NDISC: fake router advertisment received" on your
	 	hosts)
	
11/17/96	added a man page for radvd(8)

11/17/96	some minor filename changes

11/13/96	first release of this version of in.radvd (0.1)


radvd.c
/*
 *   $Id: radvd.c,v 1.1.1.1 1997/10/14 17:17:40 lf Exp $
 *
 *   Authors:
 *    Pedro Roque		<roque@di.fc.ul.pt>
 *    Lars Fenneberg		<lf@elemental.net>	 
 *
 *   This software is Copyright 1996,1997 by the above mentioned author(s), 
 *   All Rights Reserved.
 *
 *   The license which is distributed with this software in the file COPYRIGHT
 *   applies to this software. If your distribution is missing this file, you
 *   may request it from <lf@elemental.net>.
 *
 */

#include <config.h>
#include <includes.h>
#include <radvd.h>
#include <pathnames.h>

struct Interface *IfaceList = NULL;

char usage_str[] =
	"[-vh] [-d level] [-C config_file] [-m log_method] [-l log_file]\n"
	"\t[-f facility]";

#ifdef HAVE_GETOPT_LONG
struct option prog_opt[] = {
	{"debug", 1, 0, 'd'},
	{"config", 1, 0, 'C'},
	{"logfile", 1, 0, 'l'},
	{"logmethod", 1, 0, 'm'},
	{"facility", 1, 0, 'f'},
	{"version", 0, 0, 'v'},
	{"help", 0, 0, 'h'},
	{NULL, 0, 0, 0}
};
#endif

extern FILE *yyin;

char *conf_file = NULL;
char *pname;
int sock = -1;

void sighup_handler(int sig);
void timer_handler(void *data);
int readin_config(char *);
void version(void);
void usage(void);

int
main(int argc, char *argv[])
{
	unsigned char msg[MSG_SIZE];
	struct Interface *iface;
	int c, log_method;
	char *logfile;
	int facility;
#ifdef HAVE_GETOPT_LONG
	int opt_idx;
#endif

	pname = ((pname=strrchr(argv[0],'/')) != NULL)?pname+1:argv[0];

	srand((int)time(NULL));

	log_method = L_SYSLOG;
	logfile = PATH_RADVD_LOG;
	conf_file = PATH_RADVD_CONF;
	facility = LOG_FACILITY;

	/* parse args */
#ifdef HAVE_GETOPT_LONG
	while ((c = getopt_long(argc, argv, "d:C:l:m:vh", prog_opt, &opt_idx)) > 0)
#else
	while ((c = getopt(argc, argv, "d:C:l:m:vh")) > 0)
#endif
	{
		switch (c) {
		case 'C':
			conf_file = optarg;
			break;
		case 'd':
			set_debuglevel(atoi(optarg));
			break;
		case 'f':
			facility = atoi(optarg);
			break;
		case 'l':
			logfile = optarg;
			break;
		case 'm':
			if (!strcmp(optarg, "syslog"))
			{
				log_method = L_SYSLOG;
			}
			else if (!strcmp(optarg, "stderr"))
			{
				log_method = L_STDERR;
			}
			else if (!strcmp(optarg, "logfile"))
			{
				log_method = L_LOGFILE;
			}
			else if (!strcmp(optarg, "none"))
			{
				log_method = L_NONE;
			}
			else
			{
				fprintf(stderr, "%s: unknown log method: %s\n", pname, optarg);
				exit(1);
			}
			break;
		case 'v':
			version();
			break;
		case 'h':
			usage();
#ifdef HAVE_GETOPT_LONG
		case ':':
			fprintf(stderr, "%s: option %s: parameter expected\n", pname,
				prog_opt[opt_idx].name);
			exit(1);
#endif
		case '?':
			exit(1);
		}
	}
	
	if (log_open(log_method, pname, logfile, facility) < 0)
		exit(1);

	log(LOG_INFO, "version %s started", VERSION);

	/* get a raw socket for sending and receiving ICMPv6 messages */
	sock = open_icmpv6_socket();
	if (sock < 0)
		exit(1);

	/* parse config file */
	if (readin_config(conf_file) < 0)
		exit(1);

	/*
	 * okay, config file is read in, socket and stuff is setup, so
	 * lets fork now...
	 */

	if (get_debuglevel() == 0) {

		switch (fork()) {
			case 0:
				/* child falls through */
				break;
			case -1:
				log(LOG_ERR, "fork: %s", strerror(errno));
				exit(1);
			default:
				/* parent exists */
				log_close();
				exit(0);	
		}

		/*
		 * reopen logfile, so that we get the process id right in the syslog
		 */
		if (log_reopen() < 0)
			exit(1);

		/*
		 * detach from our controlling terminal
		 */
	 
		setsid();
	}

	/*
	 *	config signal handlers
	 */

	signal(SIGHUP, sighup_handler);

	/*
	 *	send initial advertisement and set timers
	 */

	for(iface=IfaceList; iface; iface=iface->next)
	{
		init_timer(&iface->tm, timer_handler, (void *) iface);
		if (iface->AdvSendAdvert)
		{
			/* send an initial advertisement */
			send_ra(sock, iface, NULL);

			set_timer(&iface->tm, iface->MaxRtrAdvInterval);
		}
	}


	/* enter loop */

	for (;;)
	{
		int len, hoplimit;
		struct sockaddr_in6 rcv_addr;
		struct in6_pktinfo *pkt_info = NULL;
		
		len = recv_rs_ra(sock, msg, &rcv_addr, &pkt_info, &hoplimit);
		if (len > 0)
			process(sock, IfaceList, msg, len, 
				&rcv_addr, pkt_info, hoplimit);
	}
}

void
timer_handler(void *data)
{
	struct Interface *iface = (struct Interface *) data;
	int next;

	dlog(LOG_DEBUG, 4, "timer_handler called for %s", iface->Name);

	send_ra(sock, iface, NULL);

	next = rand_between(iface->MinRtrAdvInterval, iface->MaxRtrAdvInterval); 
	set_timer(&iface->tm, next);
}

void
sighup_handler(int sig)
{
	struct Interface *iface;

	/* Linux has "one-shot" signals, reinstall the signal handler */
	signal(SIGHUP, sighup_handler);

	log(LOG_INFO, "attempting to reread config file");

	dlog(LOG_DEBUG, 4, "reopening log");
	if (log_reopen() < 0)
		exit(1);

	/* disable timers, free interface and prefix structures */
	for(iface=IfaceList; iface; iface=iface->next)
	{
		dlog(LOG_DEBUG, 4, "disabling timer for %s", iface->Name);
		clear_timer(&iface->tm);
	}

	iface=IfaceList; 
	while(iface)
	{
		struct Interface *next_iface = iface->next;
		struct AdvPrefix *prefix;

		dlog(LOG_DEBUG, 4, "freeing interface %s", iface->Name);
		
		prefix = iface->AdvPrefixList;
		while (prefix)
		{
			struct AdvPrefix *next_prefix = prefix->next;
			
			free(prefix);
			prefix = next_prefix;
		}
		
		free(iface);
		iface = next_iface;
	}

	IfaceList = NULL;

	/* reread config file */
	if (readin_config(conf_file) < 0)
		exit(1);

	/*
	 *	send initial advertisement and set timers
	 */
	for(iface=IfaceList; iface; iface=iface->next)
	{
		init_timer(&iface->tm, timer_handler, (void *)iface);
		if (iface->AdvSendAdvert)
		{
			/* send an initial advertisement */
			send_ra(sock, iface, NULL);
		
			set_timer(&iface->tm, iface->MaxRtrAdvInterval);
		}
	}

	log(LOG_INFO, "resuming normal operation");
}

int
readin_config(char *fname)
{
	if ((yyin = fopen(fname, "r")) == NULL)
	{
		log(LOG_ERR, "can't open %s: %s", fname, strerror(errno));
		return (-1);
	}

	if (yyparse() != 0)
		return (-1);

	return 0;
}

void
version(void)
{
	fprintf(stderr, "Version: %s\n\n", VERSION);
	fprintf(stderr, "Compiled in settings:\n", pname);
	fprintf(stderr, "  default config file		\"%s\"\n", PATH_RADVD_CONF);
	fprintf(stderr, "  default logfile		\"%s\"\n", PATH_RADVD_LOG);
	fprintf(stderr, "  default syslog facililty	%d\n", LOG_FACILITY);
#ifdef EUI64_SUPPORT
	fprintf(stderr, "  EUI-64 support		enabled\n\n"); 	
#else
	fprintf(stderr, "  EUI-64 support		disabled\n\n");
#endif	
	fprintf(stderr, "Please send bug reports or suggestions to %s.\n",
		CONTACT_EMAIL);

	exit(1);	
}

void
usage(void)
{
	fprintf(stderr,"usage: %s %s\n", pname, usage_str);
	exit(1);	
}
